File: DIP/DIP_Lab_406/Lab/Python/Lab_08/gaussian_noise_add_remove.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

def add_gaussian_noise(image, mean=0, var=0.01):
    row, col = image.shape
    sigma = var ** 0.5
    gaussian = np.random.normal(mean, sigma, (row, col))
    noisy_image = image + gaussian * 255
    return np.clip(noisy_image, 0, 255).astype(np.uint8)

def arithmetic_mean_filter(image, kernel_size=3):
    return cv2.blur(image, (kernel_size, kernel_size))

def geometric_mean_filter(image, kernel_size=3):
    image_float = image.astype(np.float32)
    img_log = np.log1p(image_float)
    kernel = np.ones((kernel_size, kernel_size), np.float32) / (kernel_size * kernel_size)
    log_filtered = cv2.filter2D(img_log, -1, kernel)
    geometric_mean_img = np.expm1(log_filtered)
    return np.clip(geometric_mean_img, 0, 255).astype(np.uint8)

# Load the image in grayscale
image = cv2.imread('lab.jpg', 0)

# Add Gaussian noise to the image
noisy_image = add_gaussian_noise(image)

# Apply arithmetic mean filter to the noisy image
arithmetic_filtered_image = arithmetic_mean_filter(noisy_image)

# Apply geometric mean filter to the noisy image
geometric_filtered_image = geometric_mean_filter(noisy_image)

# Display the images
plt.figure(figsize=(10, 5))
plt.subplot(1, 4, 1), plt.imshow(image, cmap='gray'), plt.title('Original Image')
plt.subplot(1, 4, 2), plt.imshow(noisy_image, cmap='gray'), plt.title('Noisy (Gaussian)')
plt.subplot(1, 4, 3), plt.imshow(arithmetic_filtered_image, cmap='gray'), plt.title('Arithmetic Mean Filter')
plt.subplot(1, 4, 4), plt.imshow(geometric_filtered_image, cmap='gray'), plt.title('Geometric Mean Filter')
plt.show()

File: DIP/DIP_Lab_406/Lab/353_lab_08/lab_08_MATLAB/gaussian_noise_add_remove.m
img = imread('nature.jpeg'); 
img = rgb2gray(img);  
noisy_img = imnoise(img, 'gaussian', 0, 0.01);  
arithmetic_mean_img = imfilter(noisy_img, fspecial('average', [3 3]));
[rows, cols] = size(noisy_img);
geo_img = double(noisy_img);
for i = 2:rows-1
    for j = 2:cols-1
        patch = geo_img(i-1:i+1, j-1:j+1);  
        geo_mean = exp(mean(mean(log(double(patch) + 1)))); 
        geo_img(i,j) = geo_mean - 1;  
    end
end
geo_img = uint8(geo_img);
figure;
subplot(2, 2, 1), imshow(img), title('Original Image');
subplot(2, 2, 2), imshow(noisy_img), title('Noisy (Gaussian) Image');
subplot(2, 2, 3), imshow(arithmetic_mean_img), title('Arithmetic Mean Filtered Image');
subplot(2, 2, 4), imshow(geo_img), title('Geometric Mean Filtered Image');